Cannot compile rustc on FreeBSD

Problem was introduced in 95d904625b4d45af80b4e40d51a3a0fde1abaa8a.

Command line:
```
CFG_LLVM_LINKAGE_FILE=/usr/home/aidan/rust/rust/x86_64-unknown-freebsd/rt/llvmdeps.rs \
LD_LIBRARY_PATH=/usr/home/aidan/rust/rust/x86_64-unknown-freebsd/stage1/lib:$LD_LIBRARY_PATH   \
x86_64-unknown-freebsd/stage1/bin/rustc --cfg stage1  -O --cfg rtopt --cfg debug \
-C prefer-dynamic --target=x86_64-unknown-freebsd  -D warnings \
-L "x86_64-unknown-freebsd/rt" \
-L "/usr/home/aidan/rust/rust/x86_64-unknown-freebsd/llvm/Release+Asserts/lib" \
-L ""  \
--out-dir x86_64-unknown-freebsd/stage1/lib/rustlib/x86_64-unknown-freebsd/lib \
-C extra-filename=-4e7c5e5c /usr/home/aidan/rust/rust/src/libstd/lib.rs
```
Produces error:
```
error: linking with `cc` failed: exit code: 1
note: "cc" "-L/usr/local/lib" "-L/usr/local/lib/gcc46" "-L/usr/local/lib/gcc44" "-m64" \
"-L" "/usr/home/aidan/rust/rust/x86_64-unknown-freebsd/stage1/lib/rustlib/x86_64-unknown-freebsd/lib" \
"-o" "x86_64-unknown-freebsd/stage1/lib/rustlib/x86_64-unknown-freebsd/lib/libstd-4e7c5e5c.so" \
"x86_64-unknown-freebsd/stage1/lib/rustlib/x86_64-unknown-freebsd/lib/std-4e7c5e5c.o" \
"-Wl,--whole-archive" "-lmorestack" "-Wl,--no-whole-archive" \
"x86_64-unknown-freebsd/stage1/lib/rustlib/x86_64-unknown-freebsd/lib/std-4e7c5e5c.metadata.o" \
"-nodefaultlibs" \
"/usr/home/aidan/rust/rust/x86_64-unknown-freebsd/stage1/lib/rustlib/x86_64-unknown-freebsd/lib/libcollections-4e7c5e5c.rlib" \
"/usr/home/aidan/rust/rust/x86_64-unknown-freebsd/stage1/lib/rustlib/x86_64-unknown-freebsd/lib/librand-4e7c5e5c.rlib" \
"/usr/home/aidan/rust/rust/x86_64-unknown-freebsd/stage1/lib/rustlib/x86_64-unknown-freebsd/lib/liballoc-4e7c5e5c.rlib" \
"/usr/home/aidan/rust/rust/x86_64-unknown-freebsd/stage1/lib/rustlib/x86_64-unknown-freebsd/lib/libunicode-4e7c5e5c.rlib" \
"/usr/home/aidan/rust/rust/x86_64-unknown-freebsd/stage1/lib/rustlib/x86_64-unknown-freebsd/lib/liblibc-4e7c5e5c.rlib" \
"/usr/home/aidan/rust/rust/x86_64-unknown-freebsd/stage1/lib/rustlib/x86_64-unknown-freebsd/lib/libcore-4e7c5e5c.rlib" \
"-L" "x86_64-unknown-freebsd/rt" \
"-L" "/usr/home/aidan/rust/rust/x86_64-unknown-freebsd/llvm/Release+Asserts/lib" \
"-L" "" \
"-L" "/usr/home/aidan/rust/rust/x86_64-unknown-freebsd/stage1/lib/rustlib/x86_64-unknown-freebsd/lib" \
"-L" "/usr/home/aidan/rust/rust/.rust/lib/x86_64-unknown-freebsd" \
"-L" "/usr/home/aidan/rust/rust/lib/x86_64-unknown-freebsd" \
"-Wl,--whole-archive" "-Wl,-Bstatic" "-lrust_builtin" "-lbacktrace" "-lrustrt_native" \
"-Wl,--no-whole-archive" "-Wl,-Bdynamic" "-lexecinfo" "-lpthread" "-lrt" "-lgcc_s" \
"-lpthread" "-lc" "-lm" "-shared" "-lcompiler-rt"
note: /usr/home/aidan/rust/rust/x86_64-unknown-freebsd/stage1/lib/rustlib/x86_64-unknown-freebsd/lib: file not recognized: File format not recognized

error: aborting due to previous error
/usr/home/aidan/rust/rust/mk/target.mk:166: recipe for target 'x86_64-unknown-freebsd/stage1/lib/rustlib/x86_64-unknown-freebsd/lib/stamp.std' failed
gmake: *** [x86_64-unknown-freebsd/stage1/lib/rustlib/x86_64-unknown-freebsd/lib/stamp.std] Error 101
```

The problem *seems* to be that the FreeBSD linker expects "-L<path>" to be a single argument, while `rustc` is splitting it into two arguments, "-L" and "<path>". Running the following command line by hand:

```
"cc" "-L/usr/local/lib" "-L/usr/local/lib/gcc46" "-L/usr/local/lib/gcc44" "-m64" \
"-L/usr/home/aidan/rust/rust/x86_64-unknown-freebsd/stage1/lib/rustlib/x86_64-unknown-freebsd/lib" \
"-o" "x86_64-unknown-freebsd/stage1/lib/rustlib/x86_64-unknown-freebsd/lib/libstd-4e7c5e5c.so" \
"x86_64-unknown-freebsd/stage1/lib/rustlib/x86_64-unknown-freebsd/lib/std-4e7c5e5c.o" \
"-Wl,--whole-archive" "-lmorestack" "-Wl,--no-whole-archive" \
"x86_64-unknown-freebsd/stage1/lib/rustlib/x86_64-unknown-freebsd/lib/std-4e7c5e5c.metadata.o" \
"-nodefaultlibs" \
"/usr/home/aidan/rust/rust/x86_64-unknown-freebsd/stage1/lib/rustlib/x86_64-unknown-freebsd/lib/libcollections-4e7c5e5c.rlib" \
"/usr/home/aidan/rust/rust/x86_64-unknown-freebsd/stage1/lib/rustlib/x86_64-unknown-freebsd/lib/librand-4e7c5e5c.rlib" \
"/usr/home/aidan/rust/rust/x86_64-unknown-freebsd/stage1/lib/rustlib/x86_64-unknown-freebsd/lib/liballoc-4e7c5e5c.rlib" \
"/usr/home/aidan/rust/rust/x86_64-unknown-freebsd/stage1/lib/rustlib/x86_64-unknown-freebsd/lib/libunicode-4e7c5e5c.rlib" \
"/usr/home/aidan/rust/rust/x86_64-unknown-freebsd/stage1/lib/rustlib/x86_64-unknown-freebsd/lib/liblibc-4e7c5e5c.rlib" \
"/usr/home/aidan/rust/rust/x86_64-unknown-freebsd/stage1/lib/rustlib/x86_64-unknown-freebsd/lib/libcore-4e7c5e5c.rlib" \
"-Lx86_64-unknown-freebsd/rt" \
"-L/usr/home/aidan/rust/rust/x86_64-unknown-freebsd/llvm/Release+Asserts/lib" \
"-L" \
"-L/usr/home/aidan/rust/rust/x86_64-unknown-freebsd/stage1/lib/rustlib/x86_64-unknown-freebsd/lib" \
"-L/usr/home/aidan/rust/rust/.rust/lib/x86_64-unknown-freebsd" \
"-L/usr/home/aidan/rust/rust/lib/x86_64-unknown-freebsd" "-Wl,--whole-archive" \
"-Wl,-Bstatic" "-lrust_builtin" "-lbacktrace" "-lrustrt_native" "-Wl,--no-whole-archive" \
"-Wl,-Bdynamic" "-lexecinfo" "-lpthread" "-lrt" "-lgcc_s" "-lpthread" "-lc" "-lm" \
"-shared" "-lcompiler-rt"
```
(same as before, but with `-L` and `<path>` arguments merged into a single argument)

This allowed the link to succeed.

I suspect the problem stems from the curious `-L ""`.  That's passing *two* arguments, the second of which is blank.

From my testing so far (on Linux), neither `rustc` nor `clang`/`gcc` treat `-L ""` as a synonym for `-L.`, so I'm not sure what it actually accomplishes, if anything.  `rustc` passes both arguments to `cc`.  The compiler drivers merge the two-argument form into a single-argument form for passing to `ld`.

However, Linux's `ld` actually accepts both the single-argument and two-argument forms.  When the command-line has `-L ""` on it, it corrupts the subsequent argument:
```
mkdir subdir
echo "void dummy(void){}" > subdir/dummy.c
gcc -shared subdir/dummy.c -o subdir/libdummy.so
echo  >main.rs '#[link(name = "dummy", kind = "dylib")]'
echo >>main.rs 'extern { fn dummy(); }'
echo >>main.rs 'fn main() { unsafe { dummy(); } }'
rustc main.rs -L "" -Lsubdir
```
Produces:
```
error: linking with `cc` failed: exit code: 1
note: "cc" "-Wl,--as-needed" "-m64" "-L" "/usr/local/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "./main" "./main.o" "-Wl,--whole-archive" "-lmorestack" "-Wl,--no-whole-archive" "-Wl,--gc-sections" "-pie" "-nodefaultlibs" "/usr/local/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-4e7c5e5c.rlib" "/usr/local/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcollections-4e7c5e5c.rlib" "/usr/local/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunicode-4e7c5e5c.rlib" "/usr/local/lib/rustlib/x86_64-unknown-linux-gnu/lib/librand-4e7c5e5c.rlib" "/usr/local/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-4e7c5e5c.rlib" "/usr/local/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-4e7c5e5c.rlib" "/usr/local/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-4e7c5e5c.rlib" "-L" "" "-L" "subdir" "-L" "/usr/local/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-L" "/home/rprichard/mess/dyncrate/.rust/lib/x86_64-unknown-linux-gnu" "-L" "/home/rprichard/mess/dyncrate/lib/x86_64-unknown-linux-gnu" "-Wl,--whole-archive" "-Wl,-Bstatic" "-Wl,--no-whole-archive" "-Wl,-Bdynamic" "-ldummy" "-ldl" "-lpthread" "-lrt" "-lgcc_s" "-lpthread" "-lc" "-lm" "-lcompiler-rt"
note: /usr/bin/ld: cannot find -ldummy
collect2: error: ld returned 1 exit status
```
Remove the `-L ""`, or move it to the end of the command-line, and it works.  `ld` is seeing `-L -Lsubdir`, so it adds the non-existent `-Lsubdir` directory to the search path, and it doesn't search inside `subdir`.

--- a/mk/llvm.mk
+++ b/mk/llvm.mk
@@ -44,10 +44,10 @@ $$(LLVM_STAMP_$(1)): $(S)src/rustllvm/llvm-auto-clean-trigger
 	touch -r $$@.start_time $$@ && rm $$@.start_time
 
 ifeq ($$(CFG_ENABLE_LLVM_STATIC_STDCPP),1)
-LLVM_STDCPP_LOCATION_$(1) = $$(shell $$(CC_$(1)) $$(CFG_GCCISH_CFLAGS_$(1)) \
-					-print-file-name=libstdc++.a)
+LLVM_STDCPP_RUSTFLAGS_$(1) = -L "$$(dir $$(shell $$(CC_$(1)) $$(CFG_GCCISH_CFLAGS_$(1)) \
+					-print-file-name=libstdc++.a))"
 else
-LLVM_STDCPP_LOCATION_$(1) =
+LLVM_STDCPP_RUSTFLAGS_$(1) =
 endif
 
 

--- a/mk/target.mk
+++ b/mk/target.mk
@@ -85,7 +85,7 @@ $$(TLIB$(1)_T_$(2)_H_$(3))/stamp.$(4): \
 		$$(RUST_LIB_FLAGS_ST$(1)) \
 		-L "$$(RT_OUTPUT_DIR_$(2))" \
 		-L "$$(LLVM_LIBDIR_$(2))" \
-		-L "$$(dir $$(LLVM_STDCPP_LOCATION_$(2)))" \
+		$$(LLVM_STDCPP_RUSTFLAGS_$(2)) \
 		$$(RUSTFLAGS_$(4)) \
 		--out-dir $$(@D) \
 		-C extra-filename=-$$(CFG_FILENAME_EXTRA) \

